### PR TITLE
Nokia 7210

### PR DESCRIPTION
**Pull request by hooligan-sa**
Creates device model files for:
7210-SAS-M-24F-2XFP
7210-SAS-M-1
7210-SAS-Mxp-22F2C-4SFPP
7210-SAS-T-12F10T-4XFP

### CODE REVIEW

File: device-types/Nokia/7210-SAS-M-1.yaml [hunk 1]
```diff
@@ -0,0 +1,67 @@
+---
+manufacturer: Nokia
+model: 7210 SAS-M-1
+slug: nokia-7210-sas-m-1
+part_number: 3HE04699AAAB01
+u_height: 2
+is_full_depth: false
+subdevice_role: parent
```
Comment (danner26) [inline device-types/Nokia/7210-SAS-M-1.yaml:8]: `subdevice_role: parent` is only to be used if this device is either a parent (i.e. a chassis device that houses other devices), or or a child (i.e. this device slots into a chassis device). Generally parent devices have `device-bays` defined for their children devices.

**If this is not actually a parent device, please remove this definition**

File: device-types/Nokia/7210-SAS-M-24F-2XFP.yaml [hunk 2]
```diff
@@ -0,0 +1,71 @@
+---
+manufacturer: Nokia
+model: 7210 SAS-M 24F 2XFP
+slug: nokia-7210-sas-m-24f-2xfp
+part_number: 3HE05828AAAB01
+u_height: 2
+is_full_depth: false
+subdevice_role: parent
```
Comment (danner26) [inline device-types/Nokia/7210-SAS-M-24F-2XFP.yaml:8]: `subdevice_role: parent` is only to be used if this device is either a parent (i.e. a chassis device that houses other devices), or or a child (i.e. this device slots into a chassis device). Generally parent devices have `device-bays` defined for their children devices.

**If this is not actually a parent device, please remove this definition**

File: device-types/Nokia/7210-SAS-Mxp-22F2C-4SFPP.yaml [hunk 3]
```diff
@@ -0,0 +1,75 @@
+---
+manufacturer: Nokia
+model: 7210 SAS-Mxp 22F2C 4SFPP
+slug: nokia-7210-sas-mxp-22f2c-4sfpp
+part_number: 3HE10075ABRC01
+u_height: 2
+is_full_depth: false
+subdevice_role: parent
```
Comment (danner26) [inline device-types/Nokia/7210-SAS-Mxp-22F2C-4SFPP.yaml:8]: `subdevice_role: parent` is only to be used if this device is either a parent (i.e. a chassis device that houses other devices), or or a child (i.e. this device slots into a chassis device). Generally parent devices have `device-bays` defined for their children devices.

**If this is not actually a parent device, please remove this definition**

File: device-types/Nokia/7210-SAS-T-12F10T-4XFP.yaml [hunk 4]
```diff
@@ -0,0 +1,71 @@
+---
+manufacturer: Nokia
+model: 7210 SAS-T 12F10T 4XFP
+slug: nokia-7210-sas-t-12f10t-4xfp
+part_number: 3HE08116AAAB01
+u_height: 2
+is_full_depth: false
+subdevice_role: parent
```
Comment (danner26) [inline device-types/Nokia/7210-SAS-T-12F10T-4XFP.yaml:8]: `subdevice_role: parent` is only to be used if this device is either a parent (i.e. a chassis device that houses other devices), or or a child (i.e. this device slots into a chassis device). Generally parent devices have `device-bays` defined for their children devices.

**If this is not actually a parent device, please remove this definition**

File: device-types/Nokia/7210-SAS-M-24F-2XFP.yaml [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+console-ports:
+  - name: Console
+    type: rj-45
+power-ports:
+  - name: PS1
+    type: iec-60320-c14
+    maximum_draw: 450
+  - name: PS2
+    type: iec-60320-c14
+    maximum_draw: 450
```
Comment (danner26) [inline device-types/Nokia/7210-SAS-M-24F-2XFP.yaml:18]: Is the power on these modular? Or are the power ports not removable?

File: device-types/Nokia/7210-SAS-T-12F10T-4XFP.yaml [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+console-ports:
+  - name: Console
+    type: rj-45
+power-ports:
+  - name: PS1
+    type: iec-60320-c14
+    maximum_draw: 450
+  - name: PS2
+    type: iec-60320-c14
+    maximum_draw: 450
```
Comment (danner26) [inline device-types/Nokia/7210-SAS-T-12F10T-4XFP.yaml:18]: Is the power on these modular? Or are the power ports not removable?

File: device-types/Nokia/7210-SAS-M-1.yaml [hunk 7]
```diff
@@ -0,0 +1,67 @@
+---
+manufacturer: Nokia
+model: 7210 SAS-M-1
+slug: nokia-7210-sas-m-1
+part_number: 3HE04699AAAB01
+u_height: 2
+is_full_depth: false
```
Comment (danner26) [inline device-types/Nokia/7210-SAS-M-1.yaml:7]: Please add the `airflow` attribute if you have it available

File: device-types/Nokia/7210-SAS-M-24F-2XFP.yaml [hunk 8]
```diff
@@ -0,0 +1,71 @@
+---
+manufacturer: Nokia
+model: 7210 SAS-M 24F 2XFP
+slug: nokia-7210-sas-m-24f-2xfp
+part_number: 3HE05828AAAB01
+u_height: 2
+is_full_depth: false
```
Comment (danner26) [inline device-types/Nokia/7210-SAS-M-24F-2XFP.yaml:7]: Please add the `airflow` attribute if you have it available

File: device-types/Nokia/7210-SAS-Mxp-22F2C-4SFPP.yaml [hunk 9]
```diff
@@ -0,0 +1,75 @@
+---
+manufacturer: Nokia
+model: 7210 SAS-Mxp 22F2C 4SFPP
+slug: nokia-7210-sas-mxp-22f2c-4sfpp
+part_number: 3HE10075ABRC01
+u_height: 2
+is_full_depth: false
```
Comment (danner26) [inline device-types/Nokia/7210-SAS-Mxp-22F2C-4SFPP.yaml:7]: Please add the `airflow` attribute if you have it available

File: device-types/Nokia/7210-SAS-T-12F10T-4XFP.yaml [hunk 10]
```diff
@@ -0,0 +1,71 @@
+---
+manufacturer: Nokia
+model: 7210 SAS-T 12F10T 4XFP
+slug: nokia-7210-sas-t-12f10t-4xfp
+part_number: 3HE08116AAAB01
+u_height: 2
+is_full_depth: false
```
Comment (danner26) [inline device-types/Nokia/7210-SAS-T-12F10T-4XFP.yaml:7]: Please add the `airflow` attribute if you have it available

Comment (danner26) [inline device-types/Nokia/7210-SAS-T-12F10T-4XFP.yaml:7]: Please add the `weight` and `weight_unit` attribute if you have it available

File: device-types/Nokia/7210-SAS-M-24F-2XFP.yaml (hunk 8 above)
Comment (danner26) [inline device-types/Nokia/7210-SAS-M-24F-2XFP.yaml:7]: Please add the `weight` and `weight_unit` attribute if you have it available

File: device-types/Nokia/7210-SAS-M-1.yaml (hunk 7 above)
Comment (danner26) [inline device-types/Nokia/7210-SAS-M-1.yaml:7]: Please add the `weight` and `weight_unit` attribute if you have it available

**Review by danner26 — CHANGES_REQUESTED** [review]

File: device-types/Nokia/7210-SAS-M-1.yaml [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+console-ports:
+  - name: Console
+    type: rj-45
+power-ports:
+  - name: PS1
+    type: iec-60320-c14
+    maximum_draw: 450
+  - name: PS2
+    type: iec-60320-c14
+    maximum_draw: 450
```
[the remark this answers is not in the document]
Comment (hooligan-sa) [reply device-types/Nokia/7210-SAS-M-1.yaml:18]: It's modular, but I haven't done this before. 
Do I remove the "power-ports" section and add module-bay info like below:
module-bays:
  - name: Power Supply 1
  - name: Power Supply 2
  - name: Fan Tray 1

Or do I keep both the power-ports and module-bay sections?

File: device-types/Nokia/7210-SAS-M-1.yaml (hunk 7 above)
Comment (hooligan-sa) [reply device-types/Nokia/7210-SAS-M-1.yaml:7]: Added

Comment (hooligan-sa) [reply device-types/Nokia/7210-SAS-M-1.yaml:7]: Add

File: device-types/Nokia/7210-SAS-M-1.yaml (hunk 1 above)
Comment (hooligan-sa) [reply device-types/Nokia/7210-SAS-M-1.yaml:11]: Added device-bays

File: device-types/Nokia/7210-SAS-M-24F-2XFP.yaml (hunk 8 above)
Comment (hooligan-sa) [reply device-types/Nokia/7210-SAS-M-24F-2XFP.yaml:7]: Added

File: device-types/Nokia/7210-SAS-Mxp-22F2C-4SFPP.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (hooligan-sa) [reply device-types/Nokia/7210-SAS-Mxp-22F2C-4SFPP.yaml:7]: Added

File: device-types/Nokia/7210-SAS-T-12F10T-4XFP.yaml (hunk 10 above)
Comment (hooligan-sa) [reply device-types/Nokia/7210-SAS-T-12F10T-4XFP.yaml:7]: Added

File: device-types/Nokia/7210-SAS-M-1.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (hooligan-sa) [reply device-types/Nokia/7210-SAS-M-1.yaml]: Chose to remove the "power-ports" section and add the module-bays only

File: device-types/Nokia/7210-SAS-Mxp-22F2C-4SFPP.yaml (hunk 3 above)
Comment (hooligan-sa) [reply device-types/Nokia/7210-SAS-Mxp-22F2C-4SFPP.yaml]: Removed

File: device-types/Nokia/7210-SAS-M-24F-2XFP.yaml (hunk 5 above)
Comment (hooligan-sa) [reply device-types/Nokia/7210-SAS-M-24F-2XFP.yaml]: Chose to remove the "power-ports" section and add the module-bays only

File: device-types/Nokia/7210-SAS-Mxp-22F2C-4SFPP.yaml [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+console-ports:
+  - name: Console
+    type: rj-45
+power-ports:
+  - name: PS1
+    type: iec-60320-c14
+    maximum_draw: 450
+  - name: PS2
+    type: iec-60320-c14
+    maximum_draw: 450
```
[the remark this answers is not in the document]
Comment (hooligan-sa) [reply device-types/Nokia/7210-SAS-Mxp-22F2C-4SFPP.yaml]: Chose to remove the "power-ports" section and add the module-bays only

File: device-types/Nokia/7210-SAS-M-1.yaml (hunk 1 above)
Comment (danner26) [reply device-types/Nokia/7210-SAS-M-1.yaml:11]: Which device does this house?

**Comment by danner26** [comment]
Which devices do these act as a chassis for? What is `mda2`?

**Comment by hooligan-sa** [comment]
> Which devices do these act as a chassis for? What is `mda2`?

Mea culpa, I didn't know the difference between 'device-bays' and 'module-bays' especially related to a 'subdevice_role'. 
I've corrected this and added 'mda2' as a module bay - That is the internal card type slot for the additional module that I'm aware of

File: device-types/Nokia/7210-SAS-M-1.yaml [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    type: 1000base-x-sfp
+  - name: 1/2/1
+    type: 10gbase-x-xfp
+  - name: 1/2/2
+    type: 10gbase-x-xfp
+module-bays:
+  - name: mda2
+  - name: Power Supply 1
+  - name: Power Supply 2
+  - name: Fan Tray 1
```
Comment (danner26) [inline device-types/Nokia/7210-SAS-M-1.yaml:71]: Generally you should also be adding `position` attributes to module-bays. When a module is imported in NetBox, you will need to use the position value

This is done automatically when importing a module that has its interface/port name that contains `{module}`.

File: device-types/Nokia/7210-SAS-M-24F-2XFP.yaml [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    type: 10gbase-x-xfp
+  - name: 1/2/1
+    type: 10gbase-x-xfp
+  - name: 1/2/2
+    type: 10gbase-x-xfp
+module-bays:
+  - name: mda2
+  - name: Power Supply 1
+  - name: Power Supply 2
+  - name: Fan Tray 1
```
Comment (danner26) [inline device-types/Nokia/7210-SAS-M-24F-2XFP.yaml:75]: Generally you should also be adding `position` attributes to module-bays. When a module is imported in NetBox, you will need to use the position value

This is done automatically when importing a module that has its interface/port name that contains `{module}`.

File: device-types/Nokia/7210-SAS-Mxp-22F2C-4SFPP.yaml [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+  - name: 1/1/26
+    type: 10gbase-x-sfpp
+  - name: 1/1/27
+    type: 10gbase-x-sfpp
+  - name: 1/1/28
+    type: 10gbase-x-sfpp
+module-bays:
+  - name: Power Supply 1
+  - name: Power Supply 2
+  - name: Fan Tray 1
```
Comment (danner26) [inline device-types/Nokia/7210-SAS-Mxp-22F2C-4SFPP.yaml:74]: Generally you should also be adding `position` attributes to module-bays. When a module is imported in NetBox, you will need to use the position value

This is done automatically when importing a module that has its interface/port name that contains `{module}`.

File: device-types/Nokia/7210-SAS-T-12F10T-4XFP.yaml [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+  - name: 1/1/24
+    type: 10gbase-x-xfp
+  - name: 1/1/25
+    type: 10gbase-x-xfp
+  - name: 1/1/26
+    type: 10gbase-x-xfp
+module-bays:
+  - name: Power Supply 1
+  - name: Power Supply 2
+  - name: Fan Tray 1
```
Comment (danner26) [inline device-types/Nokia/7210-SAS-T-12F10T-4XFP.yaml:70]: Generally you should also be adding `position` attributes to module-bays. When a module is imported in NetBox, you will need to use the position value

This is done automatically when importing a module that has its interface/port name that contains `{module}`.

**Review by danner26 — CHANGES_REQUESTED** [review]
Please review my other comments. Also, is the `mda2` module already in the module-types folder? If not, please create it and add that to this PR as well. We need to make sure when adding new hardware/module-bays that the assicated modules exist for importing as well

**Comment by danner26** [comment]
@hooligan-sa generally for modules we use the {module} definition in the name, so that the position value is used when placing the module

This is especially helpful when a device has 2 module bays that can have this module. It prevents interface name clashing in NetBox

**Comment by danner26** [comment]
Hello, we have released a new set of tests and features to assist with properly creating definition files for the DeviceType-Library. These tests will ensure that all files conform to a standardized format and contain valid data. Over the coming weeks we will be adding more tests to handle other currently undefined cases. If you notice anything that you believe is a bug, please @ me in a comment and details what you think is incorrect.

Along with additional test cases we have added a feature that will annotate why tests fail. To access these annotations after a test has completed, please navigate to the _Files changed_ tab within your PR. At the very bottom of the page you will see a list of errors along with the filename in question and the error which was found.

With all this being said, if your PR is validated then great! Please @ me me so I can review it further! If your PR is not valid, please review the issues and correct them; in the meantime I will mark your PR as a draft. After correcting the issues and recieving a valid status, please mark the PR as ready to be reviewed. Thank you!

**Review by danner26 — APPROVED** [review]

**Pull request merged**